python/dgl/distributed/graph_services.py

```diff
@@ -138,6 +138,30 @@ def _in_subgraph(local_g, partition_book, seed_nodes):
     global_eids = F.gather_row(local_g.edata[EID], sampled_graph.edata[EID])
     return global_src, global_dst, global_eids
 
+# --- NOTE 1 ---
+# BIG HACK AHEAD.
```

Comment:
Could you rewrite it in as a formal docstring. Right now it reads like a casual note for yourself.
- Instead of claiming it as a hack, it is more like a current limitation to me.
- Avoid "I do ...". Use "The implementation does ...".
- "We do ..." is ok but should be used with caution. For example, "If we only use edge data, everything will be fine because the local edge data ..." sounds very casual. Consider change it to "The current implementation only works for sampling algorithms that rely on the data of the seed nodes and neighboring edges because they are locally available. It is not applicable to algorithms that rely on the data of neighboring nodes such as ...".
- "If we don't want to overhaul the design of DistDGL, the best thing I can think of is to cache the halo node features, which (1) incurs a lot more extra memory cost, and (2) needs significant refactor on DistGraph and/or DistTensor." -> This should belong to a design doc or an issue. The general audience will not fully understand this.

Think of you are writing a paper or essay. You will never use some of the tones here. The idea is the same for docstring/comment.

Reply:
I also think what you described here is the general design philosophy (or restriction) of the entire `graph_services.py` so you could put it to the beginning of file.